src/v/raft/consensus.cc

```diff
@@ -1242,23 +1242,84 @@ consensus::remove_member(vnode node, model::revision_id new_revision) {
       });
 }
 
+ss::future<std::optional<model::offset>>
+consensus::adjust_learner_initial_offset(
+  std::optional<model::offset> learner_start_offset) {
+    if (!learner_start_offset) {
+        co_return learner_start_offset;
+    }
+    /**
+     * We need to adjust the offset to the full batch boundary. This
+     * is required as the last offset included in the snapshot MUST be
+     * the last offset in a batch. If that wouldn't be the case Raft
+     * leader would not be able to recover follower as Raft delivers
+     * whole batches not individual records. We are looking for the last offset
+     * of a batch that is smaller than the requested learner_start_offset.
+     *
+     *
+     * The following invariant MUST hold:
+     * adjusted_offset <= last_included_offset
+     */
+    storage::log_reader_config rdr_config(
+      *learner_start_offset,
+      model::offset::max(),
+      _scheduling.default_iopc,
+      _as);
+    rdr_config.skip_batch_cache = true;
+    rdr_config.max_bytes = 1;
+
+    auto rdr = co_await _log->make_reader(rdr_config);
```

Comment:
Hmm, we should be able to get what we need without actually reading just by replicating the storage logic there, right?

In fact I think `disk_log_impl::index_lower_bound` implementation has a bug - usually it returns max last offset that is strictly less than the needle, but if `idx.max_offset() == o` it returns last offset that is equal (and not strictly less as in other cases).

This bug aside, I think what we need here doesn't require any prev/next offsets - just do `segments().lower_bound(o)` and then `it->idx.find_nearest(o)`. It will return what we need - max base offset that is less or equal to the needle.